### PR TITLE
feat(files): adds download permission checks

### PR DESCRIPTION
**Pull request by hypeJunction**
File download permissions can now be filtered via hooks.

### CODE REVIEW

File: engine/classes/Elgg/UserCapabilities.php [hunk 1]
```diff
@@ -98,21 +99,25 @@ public function canEdit(ElggEntity $entity, $user_guid = 0) {
...
+		];
+
 		return $this->hooks->trigger('permissions_check', $entity->getType(), $params, $default);
 	}
 
 	/**
 	 * Can a user delete this entity?
 	 *
-	 * @tip Can be overridden by registering for the permissions_check:delete plugin hook.
+	 * @tip   Can be overridden by registering for the permissions_check:delete plugin hook.
```
Comment (jdalsem) [inline engine/classes/Elgg/UserCapabilities.php]: Not needed

File: engine/classes/Elgg/UserCapabilities.php [hunk 2]
```diff
@@ -98,21 +99,25 @@ public function canEdit(ElggEntity $entity, $user_guid = 0) {
...
 	 *
 	 * @param ElggEntity $entity    Object entity
 	 * @param int        $user_guid The user GUID, optionally (default: logged in user)
 	 *
 	 * @return bool Whether this entity is deletable by the given user.
 	 * @since 1.11
-	 * @see elgg_set_ignore_access()
+	 * @see   elgg_set_ignore_access()
```
Comment (jdalsem) [inline engine/classes/Elgg/UserCapabilities.php]: Not needed

File: engine/classes/Elgg/UserCapabilities.php [hunk 3]
```diff
@@ -127,22 +132,23 @@ public function canDelete(ElggEntity $entity, $user_guid = 0) {
...
 		];
+
 		return $this->hooks->trigger('permissions_check:delete', $entity->getType(), $params, $return);
 	}
 
 	/**
 	 * Can a user delete this river item?
 	 *
-	 * @tip Can be overridden by registering for the "permissions_check:delete", "river" plugin hook.
+	 * @tip   Can be overridden by registering for the "permissions_check:delete", "river" plugin hook.
```
Comment (jdalsem) [inline engine/classes/Elgg/UserCapabilities.php]: Not needed

File: engine/classes/Elgg/UserCapabilities.php [hunk 4]
```diff
@@ -127,22 +132,23 @@ public function canDelete(ElggEntity $entity, $user_guid = 0) {
...
 	 *
-	 * @note This is not called by elgg_delete_river().
+	 * @note  This is not called by elgg_delete_river().
```
Comment (jdalsem) [inline engine/classes/Elgg/UserCapabilities.php]: Not needed

File: engine/classes/Elgg/UserCapabilities.php [hunk 5]
```diff
@@ -127,22 +132,23 @@ public function canDelete(ElggEntity $entity, $user_guid = 0) {
...
 	 *
 	 * @param ElggRiverItem $item      River item
 	 * @param int           $user_guid The user GUID, optionally (default: logged in user)
 	 *
 	 * @return bool Whether this river item should be considered deletable by the given user.
 	 * @since 2.3
-	 * @see elgg_set_ignore_access()
+	 * @see   elgg_set_ignore_access()
```
Comment (jdalsem) [inline engine/classes/Elgg/UserCapabilities.php]: Not needed

File: engine/classes/ElggFile.php [hunk 6]
```diff
@@ -469,4 +469,66 @@ public function __sleep() {
...
+
+		$file_svc = new \Elgg\FileService\File();
+		$file_svc->setFile($this);
+		$file_svc->setExpires($expires);
+		$file_svc->setDisposition('attachment');
+		$file_svc->bindSession($use_cookie);
+		$url = $file_svc->getURL();
+
+		$params = [
+			'entity' => $this,
```
Comment (jdalsem) [inline engine/classes/ElggFile.php]: Should we add the file_svc to the params? Makes changing the url a bit easier

File: engine/classes/ElggFile.php [hunk 7]
```diff
@@ -469,4 +469,66 @@ public function __sleep() {
...
+		if ($expires) {
+			$file_svc->setExpires($expires);
+		}
+		$file_svc->setDisposition('inline');
+		$file_svc->bindSession($use_cookie);
+		$url = $file_svc->getURL();
+
+		$params = [
+			'entity' => $this,
+		];
```
Comment (jdalsem) [inline engine/classes/ElggFile.php]: add file_svc?

**Review by jdalsem — APPROVED** [review]
Besides my comments this LGTM

File: engine/classes/ElggFile.php (hunk 6 above)
Comment (hypeJunction) [inline engine/classes/ElggFile.php]: I don't think anyone should be mendling with file_svc

Comment (jdalsem) [inline engine/classes/ElggFile.php]: if you want to replace the url with something else (not cookie bound, or another timeout) you will need the file_svc anyway. In that case you can create one again...

Comment (hypeJunction) [inline engine/classes/ElggFile.php]: `return $file->getDownloadURL(false)`

Comment (hypeJunction) [inline engine/classes/ElggFile.php]: No, don't like it. File service is bound to instance of ElggFile and other parameters of the download/inline URL. I'd rather plugins instantiate it again and build their own parameters if that's an intention. My motivation is to fallback to non /serve-file/ URLs if needed. 
You can add the file service later if there is a compelling case and add the appropriate tests.

**Pull request merged**